### DISCUSSION TITLE
Unable to create backup: unexpected failure invoking barman-cloud-wal-archive: exit status 4

### DISCUSSION THREAD

**Discussion by AllardKrings** [cloudnative-pg/cloudnative-pg#4376] [Q&A]
hi,

i am running cnpg in combination with minio on an arm sbc with ubuntu 23.10 and microk8s 1.29.

my cluster is defined as:

apiVersion: postgresql.cnpg.io/v1
kind: Cluster
metadata:
name: postgres13
namespace: postgres
spec:
instances: 3
imageName: ghcr.io/cloudnative-pg/postgresql:13.14-3
bootstrap:
initdb:
postInitSQL:
- CREATE USER deptrack WITH PASSWORD 'deptrack'
- CREATE DATABASE deptrack OWNER deptrack
storage:
size: 5Gi
backup:
barmanObjectStore:
destinationPath: 's3://backups/'
endpointURL: 'http://minio.postgres:9000/'
s3Credentials:
accessKeyId:
name: minio-creds
key: MINIO_ACCESS_KEY
secretAccessKey:
name: minio-creds
key: MINIO_SECRET_KEY

Minio runs in the same namespace:

NAME READY STATUS RESTARTS AGE
minio-547b5c995b-6dwm5 1/1 Running 2 (5h28m ago) 25h
postgres13-1 1/1 Running 0 3h20m
postgres13-2 1/1 Running 0 3h19m
postgres13-3 1/1 Running 0 3h19m

I creates a bucket “backups” in minio.

When I run a backup:

apiVersion: postgresql.cnpg.io/v1
kind: Backup
metadata:
name: postgres13-backup
namespace: postgres
spec:
cluster:
name: postgres13

It gives an error:

Name: postgres13-backup
Namespace: postgres
Labels:
Annotations:
API Version: postgresql.cnpg.io/v1
Kind: Backup
Metadata:
Creation Timestamp: 2024-04-25T07:29:24Z
Generation: 1
Resource Version: 3755479
UID: ea5ef0c7-098b-4c23-870a-d63b8c65a63b
Spec:
Cluster:
Name: postgres13
Method: barmanObjectStore
Status:
Backup Name: backup-20240425072926
Destination Path: s3://backups/
Endpoint URL: http://minio.postgres:9000/
Instance ID:
Container ID: containerd://86707c09825439d602b3200030f313be4e42d02a461bbfc10bea501900058573
Pod Name: postgres13-2
Method: barmanObjectStore
Phase: walArchivingFailing
s3Credentials:
Access Key Id:
Key: MINIO_ACCESS_KEY
Name: minio-creds
Secret Access Key:
Key: MINIO_SECRET_KEY
Name: minio-creds
Server Name: postgres13
Events:

The log of postgres13-1 pod says:

{"level":"error","ts":"2024-04-25T10:50:09Z","logger":"wal-archive","msg":"failed to run wal-archive command","logging_pod":"postgres13-1","error":"unexpected failure invoking barman-cloud-wal-archive: exit status 4","stacktrace":"github.com/cloudnative-pg/cloudnative-pg/pkg/management/log.(*logger).Error\n\tpkg/management/log/log.go:128\ngithub.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/walarchive.NewCmd.func1\n\tinternal/cmd/manager/walarchive/cmd.go:95\ngithub.com/spf13/cobra.(*Command).execute\n\tpkg/mod/github.com/spf13/cobra@v1.8.0/command.go:983\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\tpkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1115\ngithub.com/spf13/cobra.(*Command).Execute\n\tpkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1039\nmain.main\n\tcmd/manager/main.go:64\nruntime.main\n\t/opt/hostedtoolcache/go/1.21.5/x64/src/runtime/proc.go:267"}
{"level":"info","ts":"2024-04-25T10:50:09Z","logger":"postgres","msg":"record","logging_pod":"postgres13-1","record":{"log_time":"2024-04-25 10:50:09.867 UTC","process_id":"29","session_id":"662a056c.1d","session_line_num":"759","session_start_time":"2024-04-25 07:25:32 UTC","transaction_id":"0","error_severity":"LOG","sql_state_code":"00000","message":"archive command failed with exit code 1","detail":"The failed archive command was: /controller/manager wal-archive --log-destination /controller/log/postgres.json pg_wal/000000010000000000000001","backend_type":"archiver"}}
{"level":"info","ts":"2024-04-25T10:50:09Z","logger":"postgres","msg":"record","logging_pod":"postgres13-1","record":{"log_time":"2024-04-25 10:50:09.867 UTC","process_id":"29","session_id":"662a056c.1d","session_line_num":"760","session_start_time":"2024-04-25 07:25:32 UTC","transaction_id":"0","error_severity":"WARNING","sql_state_code":"01000","message":"archiving write-ahead log file "000000010000000000000001" failed too many times, will try again later","backend_type":"archiver"}}

What am I doing wrong?

Help apprectated!

**Comment by 8ball030** [comment]
same issue here

**Comment by ololdach** [comment]
Same here. In my case it was the disk space in the container/PVC that prevented the backup file to be successfully created

**Comment by matlawsky** [comment]
Did anyone solve this issue?

**Comment by npsables** [comment]
same issue here

**Comment by djeinstine** [comment]
Same issue. Running postgres 15.8

**Comment by zeddit** [comment]
+1

**Comment by acsezen** [comment]
Same here...

**Comment by hbobenicio** [comment]
By looking at this source code (https://github.com/EnterpriseDB/barman/blob/master/barman/clients/cloud_walarchive.py) which I suppose is the implementation of the command which is failing here (not really sure, sorry. If I'm mistaken, please someone correct me), we can see there is a GeneralErrorExit exception class that maps to the exit code 4. The exception itself doesn't seem to have much context of the root exception, but there're logs right above it's instantiation/raising. Maybe checking the logging right before this error log may contain an exception which may be the root cause of the problem. Maybe greping something like 'Barman cloud WAL archiver exception' or 'Exception details' can give more insights about what's going on

**Comment by acsezen** [comment]
Is the target location empty ? Does it truncated before additional installation attempts?

**Comment by bw210343** [comment]
For me this error occurred when I enabled encryption on the back up files

```
  wal:
    # -- Whether to instruct the storage provider to encrypt WAL files. One of `` (use the storage container default), `AES256` or `aws:kms`.
    encryption: AES256
```

Removing the encryption resolved the issue:
```
  wal:
    # -- Whether to instruct the storage provider to encrypt WAL files. One of `` (use the storage container default), `AES256` or `aws:kms`.
    encryption: 
```

**Comment by Adrelien** [comment]
Having same issue

**Comment by mzupan** [comment]
so I got around this. I did the data without encryption to use bucket encryption also.

```
  wal:
    encryption: ""
  data:
    encryption: ""
```

Also i tried to use pod-identity and that doesn't work. My guess is the version of boto or something is too old they are using. I had to use the annotation for the service account. 

Also the chart doesn't support not using key/secret so i forked it to make

```
  {{- $secretName := coalesce .scope.secret.name (printf "%s-%s-s3-creds" .chartFullname .secretPrefix) }}
  s3Credentials:
    inheritFromIAMRole: true
```

An option in the chart but i might send a PR for this so it works upstream also

Also if you add this after cluster is created it seems you need to restart the replica for it to work

**Comment by bdsoha** [comment]
Still not working for me

**Comment by hamedmosavat** [comment]
My problem was that there were already some WALs in the object store with the same name, like 0000000100000000. After I make it empty, it got resolved. So, it was not able to write new WALs on the old ones.

**Reply by dshmmd** [reply]
thank you! it worked.

**Reply by Atomique** [reply]
Thanks a lot. I tried to experiment with cnpg and wanted to try a restore on a cluster. Problem was that I deleted the source cluster and recreated a cluster with the same name. Wanted to something like a desaster recovery. After renaming my "new" cluster that restores the source cluster my problem was solved.

**Reply by hgyowan** [reply]
thank you! it worked.

**Comment by majermarci** [comment]
For me the `15.10` image had issues like this, and some others that were released probably around that (like 14.15).

Solution was to use a more recent image of this version, as the 15.10 I had on my cluster is most probably buggy and pull policy was on `IfNotPresent`.

Now it works with image `15.10-33-bookworm@sha256:542eddeb15a4b7cf80eebabf716a9bb1ade9700702eea2e226645d19a60606fe` as an example.

**Comment by badchoyo** [comment]
I found a similar error in this version - ghcr.io/cloudnative-pg/postgresql:17.5
downgraded the version to imageName: ghcr.io/cloudnative-pg/postgresql:16.3
the error with WAL is gone.

**Comment by hafizjamshed** [comment]
I found the similar problem in my backup exit code 4. 
barman-cloud-backup uses a different name when took the backup and cnpg is asking for the list with a different name. 
I did the manual backup and my backup is there hetzner cloud. 
Can someone guide me, how to proceed with that. I am newbie {"level":"error","ts":"2025-08-21T14:14:28.521433684Z","logger":"barman","msg":"Can't extract backup id","backupName":"postgresql-cluster-20250821161419","backupNamespace":"cnpg-system","logging_pod":"postgresql-cluster-1","command":"barman-cloud-backup-show","options":["--format","json","--endpoint-url","https://ha-kubernetes.nbg1.your-objectstorage.com","--cloud-provider","aws-s3","s3://database-backup","postgresql-cluster","backup-20250821141419"],"stdout":"","stderr":"2025-08-21 14:14:28,443 [432] ERROR: Barman cloud backup show exception: Unknown backup 'backup-20250821141419' for server 'postgresql-cluster'\n","error":"exit status 4","stacktrace":"github.com/cloudnative-pg/machinery/pkg/log.(*logger).Error\n\tpkg/mod/github.com/cloudnative-pg/machinery@v0.3.1/pkg/log/log.go:125\ngithub.com/cloudnative-pg/barman-cloud/pkg/command.executeQueryCommand\n\tpkg/mod/github.com/cloudnative-pg/barman-cloud@v0.3.3/pkg/command/backuplist.go:88\ngithub.com/cloudnative-pg/barman-cloud/pkg/command.GetBackupByName\n\tpkg/mod/github.com/cloudnative-pg/barman-cloud@v0.3.3/pkg/command/backuplist.go:141\ngithub.com/cloudnative-pg/barman-cloud/pkg/backup.(*Command).GetExecutedBackupInfo\n\tpkg/mod/github.com/cloudnative-pg/barman-cloud@v0.3.3/pkg/backup/backup.go:138\ngithub.com/cloudnative-pg/cloudnative-pg/pkg/management/postgres.(*BackupCommand).takeBackup\n\tpkg/management/postgres/backup.go:204\ngithub.com/cloudnative-pg/cloudnative-pg/pkg/management/postgres.(*BackupCommand).run\n\tpkg/management/postgres/backup.go:156"}
Barman-cloud stores the backup with this name 20250821T141419 And CNPG is asking 20250821141419.

**Reply by Biskwit** [reply]
Same here, the verification fail but the backup is present on s3, what's your operator version ?
